SCPCP000006 - Add 01-clustering template analysis

maud-p opened the pull request:
### Purpose/implementation Section


#### Please link to the GitHub issue that this pull request addresses.

https://github.com/AlexsLemonade/OpenScPCA-analysis/discussions/635

#### What is the goal of this pull request?

As discussed in the previous PR (https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/699) I wanted to have one notebook that is rendered over the entire Wilms tumor dataset.


#### Briefly describe the general approach you took to achieve this goal.

1. I basically took the previous .Rmd file, 
2. remove all the comments specific to one sample, 
3. remove the label transfer using the fetal kidney atlas (for the next PR)
4. write a R script to render the notebook on all samples. 

#### If known, do you anticipate filing additional pull requests to complete this analysis module?

Yes! More than one. At least:

- [ ] A PR to build the fetal kidney reference for Azimuth,
- [ ] A PR to include the label transfer from the fetal kidney reference to the clustering template.
- [ ] A PR to include inferCNV and/or copyKAT to the template.
- [ ] A PR to subdivide and re-cluster stroma and fetal nephron cells, taht contain nromal + cancer cells


### Results

#### What types of results does your code produce (e.g., table, figure)?

HTML reports 

#### What is your summary of the results?




### Provide directions for reviewers

_In this section, tell reviewers what kind of feedback you are looking for._
_This information will help guide their review._

#### What are the software and computational requirements needed to be able to run the code in this PR?

_This information will help reviewers run the code during review, if applicable._
_For software, how should reviewers set up their environment (e.g., `renv` or `conda`) to run this code?_
_For compute, can reviewers run this code on their laptop, or do they need additional computational resources such as RAM or storage?_
_Please make sure this information, if applicable, is documented in the README.md file._



#### Are there particularly areas you'd like reviewers to have a close look at?



#### Is there anything that you want to discuss further?

If you have any comments, ideas, very welcome. Either on the content of the template (as I said, I think some sections will not be so usefull) or the future plans:
- [ ] build the fetal kidney reference for Azimuth,
- [ ] include the label transfer from the fetal kidney reference to the clustering template.
- [ ] include inferCNV and/or copyKAT to the template.
- [ ] subdivide and re-cluster stroma and fetal nephron cells, taht contain nromal + cancer cells




### Author checklists

#### Analysis module and review

- [x] This analysis module [uses the analysis template and has the expected directory structure](https://openscpca.readthedocs.io/en/latest/contributing-to-analyses/analysis-modules/).
- [x] The analysis module `README.md` has been updated to reflect code changes in this pull request.
- [x] The analytical code is documented and contains comments.
- [ ] Any results and/or plots this code produces have been added to your S3 bucket for review.

#### Reproducibility checklist

- [ ] Code in this pull request has been added to the GitHub Action workflow that runs this module.
- [x] The dependencies required to run the code in this pull request have been added to the analysis module `Dockerfile`.
- [ ] If applicable, the dependencies required to run the code in this pull request have been added to the analysis module conda `environment.yml` file.
- [x] If applicable, R package dependencies required to run the code in this pull request have been added to the analysis module `renv.lock` file.

sjspielman commented:
Hi @maud-p, thanks for filing this updated PR! I wanted to let you know that I'm going over your code now, and you can expect a review from me tomorrow.

sjspielman commented on analyses/cell-type-wilms-tumor-06/marker-sets/Pathway_metadata.csv:
This file can be deleted (see overall comment).

sjspielman commented on analyses/cell-type-wilms-tumor-06/render_01-clustering.R:
```diff
@@ -0,0 +1,92 @@
+  
+  rmarkdown::render(input = "notebook-template/01-clustering.Rmd", 
+                    output_format = "html_document",
+                    output_file = paste0("01-clustering_",i,"_",gsub(paste0(data_dir, "/"), "", filelist[i]), ".html"),
```
We don't need to keep this `i` index in the file name (note that you may need to remove old `html` versions when you re-render that do have the `_i_`).

```suggestion
                    output_file = paste0("01-clustering_",gsub(paste0(data_dir, "/"), "", filelist[i]), ".html"),
```

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+---
+title: "Evaluation of clustering parameters"
```
This notebook isn't exactly evaluating _parameters_, since you fixed those from the beginning. Instead, how about this?

```suggestion
title: "Exploration of clustering results"
```

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+    code_folding: hide
+    highlight: pygments
+    df_print: kable
+params:
+  sample_name: ~
+  sample_path: ~
+  out_rds_path: ~
```
As written in my overall comment, you'll want to use some actual defaults there that will allow the notebook to run.

maud-p replied:
thanks, I didn't know about the params, I'll try to use it starting from now!

sjspielman replied:
They're super useful! Here's some info about them 🙂:
- https://docs.posit.co/connect/user/param-rmarkdown/
- https://rmarkdown.rstudio.com/lesson-6.html

maud-p replied:
Thank you!

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+
+## Introduction
+
+This a template for an analysis notebook using RMarkdown.
```
```suggestion
```

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+- https://github.com/AlexsLemonade/OpenScPCA-analysis/discussions/635
+- https://github.com/AlexsLemonade/OpenScPCA-analysis/issues/671
+
+In order to assess the clustering quality, we look into some marker genes, pathways enrichment and label transfer.
```
Again, I would argue that you're not really assessing quality (that would involve an in-depth investigation of clustering parameter values and assessing robustness of clustering to different parameters), but more exploring results from clustering, so I might rephrase this a smidge.

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+- FindClusters (default)
+
+
+Note: Other parameters have been previously tested, but we would like to show in the following report that the one selected is performing good.
```
Unless you are prepared to show the full set of analyses that led you to use these parameters, I would not include this. 
```suggestion
```

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+# convert to seurat
+srat <- CreateSeuratObject(counts = counts(sce),
+                                    assay = "RNA",
+                                    project = sample
```
This is an example of where you would use a param. If you have a `sample_id` parameter defined, this would become:

```suggestion
                                    project = params$sample_id
```

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+
+
+# Normalization
+options(future.globals.maxSize= 8912896000000)
```
This is an example of a "setup" line of code - let's move this to the top of the notebook around the chunk you'll add which loads libraries.

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+
+# add metadata from SingleCellExperiment to Seurat
+srat@misc <- metadata(sce)
+
```
At this point, I recommend deleting the sce to save some space/memory, since after this point we're in Seurat land

```suggestion
rm(sce)
```

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+
+Please note: to keep the notebook as straight as possible, we decided to show the analysis for the selected set of parameters:
+
+- SCTransform (default parameters)
+- RunPCA (default) 
+- RunUMAP (dims = 1:50) 
+- FindNeighbors (dims = 1:50) 
+- FindClusters (default)
```
Example (see my overall comment) of what I mean to use backticks for code-y things:

```suggestion
- `SCTransform` (default parameters)
- `RunPCA` (default) 
- `RunUMAP` (dims = 1:50) 
- `FindNeighbors` (dims = 1:50) 
- `FindClusters` (default)
```

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
````diff
@@ -0,0 +1,768 @@
+
+## Sample metadata
+
+```{r paths2}
+## to be modified when rendered
+filedir <- filelist[i] 
+sample <- gsub(paste0(data_dir, "/"), "", filedir)
+
+metadata <- read.table(file.path(data_dir, "/single_cell_metadata.tsv"), sep = "\t", header=TRUE)
+
+metadata[metadata$scpca_sample_id == sample,]
+```
````
This entire chunk can be deleted, since you will be using a different strategy to read in files, and we don't need this `metadata` data frame.

sjspielman commented on analyses/cell-type-wilms-tumor-06/notebook-template/01-clustering.Rmd:
```diff
@@ -0,0 +1,768 @@
+This part is imho one of the most important step that allow us to have a quick and reliable idea of the composition of the different clusters. 
+The predicted compartment are defined into 4 categories:
+
+- endotheliúm
```
```suggestion
- endothelium
```

sjspielman commented on analyses/cell-type-wilms-tumor-06/render_01-clustering.R:
```diff
@@ -0,0 +1,92 @@
+repository_base <- rprojroot::find_root(rprojroot::is_git_root)
+
+# The current data directory, found within the repository base directory
+data_dir <- file.path(repository_base, "data", "2024-07-08", "SCPCP000006")
```
We'd like to use `current` here FYI, not the date; see the "tip" here: https://openscpca.readthedocs.io/en/latest/getting-started/accessing-resources/getting-access-to-data/#downloaded-data-file-structure

```suggestion
data_dir <- file.path(repository_base, "data", "current", "SCPCP000006")
```

sjspielman commented:
I want to also note that this comment https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/706#discussion_r1711482392 @jaclyn-taroni's review in your other PR could very much also apply to this notebook too. Analysis parameter choices, which you may decide to vary at some future point, are a good type of variable to make into an R Markdown parameter.

maud-p commented:
Thank you very much @sjspielman for taking the time to give such a great feedback!

I'll take some more time to read all your documentation point by point but globally from a first read it is all clear and make lot of sense!

To answer your question about the sensitivity of DElegate to the number of clusters, I think it is!
Naively, I used it here to have a rapid idea of the quality of the clustering: if it finds specific genes for each cluster, I expect each of them to have at least a different phenotype. In case it struggles finding specific genes, we might have over-clustered.
It would also allow a quick check if our "favorite" known marker genes pop up in the top 5, like CD45.

As you wrote, I would not use it for annotation. The two uses of it I can see would be (i) at the very beginning to get an idea of the dataset, and (ii) at the very end of the analysis module to find marker genes of the annotated cell types. Would that makes sense?

For plannification, I'd like to first fix the PR #706 regarding the Azimuth/Seurat reference for label transfer and then focus on this PR. 

I'll try to provide you feedback hopefully by the end of next week!

Thank you again!

maud-p commented:
Thank you very much again @sjspielman for your detailed review and all comments!

I tried to improve the script and RMarkdown following your advice. Main points:

-	I shortened the R script to the `for-loop` and moved the rest to the RMarkdown following the template.
-	I used the `params` in the `yaml` section of the RMarkdown
-	I defined `functions` for repetitive codes. 
-	I placed the R script in `/script` folder, the RMarkdown template in the `template` folder and the notebooks in the `notebook` folder. I added a short `READ.ME` file in the `template folder.

FYI, few more changes are planed with PR #706. Not sure what is best to do first, so I just decided to inform you about the few changes I have made to this PR!

Thank you again,

sjspielman commented:
> FYI, few more changes are planed with PR https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/706. Not sure what is best to do first, so I just decided to inform you about the few changes I have made to this PR!

I was just checking in with some of the changes and @jaclyn-taroni's feedback in #706 when you posted this comment, so good timing 😄 ! I'll have a look at the changes in this PR today, but I do have a sense that some of the functions you are going to write as part of #706 will impact this code, too. 

I think a first step towards holistically approaching both PRs would be to first "inventory" the code and figure out which functions you are going to write (e.g., function(s) for common Seurat steps). This will help us understand which functions being written in #706 should also be used here. This leads into some logistics we'll want to keep track of to make sure you don't double your work, and help you avoid conflicts in your code! Specifically, if there are functions which are going to be used in _both_ PRs, then you'll want to tackle one PR fully at a time.

Because I anticipate you'll be writing functions as part of #706 that will also be used in this notebook, I'd suggest first working to wrap up #706. Once that is merged into `main`, we can update this branch (`01-clustering-template`) to sync changes, allowing us to use code in #706 in this PR directly (see our [documentation on syncing](https://openscpca.readthedocs.io/en/latest/contributing-to-analyses/working-with-git/staying-in-sync-with-upstream/) - this can be tricky so please let me know how I can help with this more directly when we get there!). In the meantime, I'll leave feedback today on the code changes you've made so far, but I would probably recommend focusing on #706 before coming back to this PR.

maud-p commented:
Great thank you! I agree to first focus on PR #706 and then coming back to this one 😄

maud-p commented on analyses/cell-type-wilms-tumor-06/scripts/render_01-clustering.R:
```diff
@@ -0,0 +1,16 @@
+sample_metadata_file <- file.path(root_dir, "data", "current", "SCPCP000006", "single_cell_metadata.tsv")
+metadata <- read.table(sample_metadata_file, sep = "\t", header = TRUE)
+
+for (i in metadata$scpca_sample_id[9:11]) {
```
Yes it was for testing! I was running the loop from the first sample and realized I could test on small samples to check it is running (9 to 11 are small)

maud-p commented:
Hi @sjspielman ,
I just wanted to let you know that I will continue working on the cluster exploration now 😄 
I'll work on a branch from https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/737 once it will be merged and open a new PR once I have the script updated! Hopefully early next week 😄

maud-p commented:
I think we can close this PR then.

sjspielman commented:
Hi @maud-p, if you think a new PR makes more sense given the other module changes you've been making, all good with me! Let's leave this PR open until a new one is filed, just in case.

Either way, I'm not sure if you've had a chance to encounter this yet, but you might find _cherry-picking_ git commits useful as you handle logistics here. Cherry-picking allows you to literally duplicate a commit between branches. For example, if there are commits in this branch you want to include in a different branch to ultimately file a new PR, you can use cherry-picking to copy them over. Here's a tutorial on using it with GitKraken & via command line! https://www.gitkraken.com/learn/git/cherry-pick 

Feel free to let me know (on Github or Slack!) if I can help with any git things along the way here :)

maud-p commented:
Thank you very much @sjspielman , sounds very useful and I didn't know about it - actually you can consider I don't know anything about `git`, learning right now :)

sjspielman commented:
> actually you can consider I don't know anything about git, learning right now :)

You're doing great!! It's a lot to learn..always a journey!
